cow_instrument/testing/IndexTranslatorTest.cpp

```diff
@@ -0,0 +1,229 @@
+  EXPECT_EQ(spectrumNumbers.size(), 3);
+  EXPECT_EQ(spectrumNumbers[0], std::vector<SpectrumNumber>({5}));
+  EXPECT_EQ(spectrumNumbers[1], std::vector<SpectrumNumber>({2}));
+  EXPECT_EQ(spectrumNumbers[2], std::vector<SpectrumNumber>({}));
+}
+
+TEST(IndexTranslatorTest, test_split_2_to_3_combine_results) {
+  // This tests mimics what would happen in an MPI redistribution
+  Partitioning p2(2);
+  std::vector<IndexTranslator> t;
+  t.emplace_back(
+      p2, 0, std::vector<SpectrumNumber>{1, 2, 4, 5, 6, 8, 9}); // partition 0
+  t.emplace_back(
+      p2, 1, std::vector<SpectrumNumber>{1, 2, 4, 5, 6, 8, 9}); // partition 1
+  Partitioning p3(3);
+
+  // Fake data object, could be anything.
+  using Data = size_t;
+
+  // Outermost vector is for target partitioning
+  // Middle vector is for input partitioning
+  // Innermost vector is spectrum index of input
+  std::vector<std::vector<
+      std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>>>>
+      receiveBuffer(p3.size());
+
+  // This for-loop mimics MPI ranks *before* redistribution
+  for (const auto partition : {0, 1}) {
+    // 1. Get local spectrum indices (used to extract data)
+    auto indices = t[partition].spectrumIndices(p3);
+
+    // 2. Use indices to get data from object that should be redistributed. Here
+    // we simply keep indices as fake data.
+    const auto &data = indices;
+
+    // 3. Tag data with GlobalSpectrumIndex and SpectrumNumber
+    auto globalIndices = t[partition].globalSpectrumIndices(p3);
+    auto spectrumNumbers = t[partition].spectrumNumbers(p3);
+    std::vector<
+        std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>>>
+        sendBuffer(p3.size());
+    // Loop over target partition
+    for (size_t p = 0; p < indices.size(); ++p) {
+      for (size_t i = 0; i < indices[p].size(); ++i) {
+        sendBuffer[p].emplace_back(globalIndices[p][i], spectrumNumbers[p][i],
+                                   data[p][i]);
+      }
+    }
+
+    // 4. Send data, e.g., with MPI_Isend(). Here we fake a send by simply
+    // appending the receive buffer.
+    for (size_t p = 0; p < indices.size(); ++p)
+      receiveBuffer[p].push_back(sendBuffer[p]);
+  }
+
+  // For-loop mimic MPI ranks *after* redistribution
+  for (const auto partition : {0, 1, 2}) {
+    // 5. Merge data from all ranks
+    std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>> result;
+    // Loop over initial partition
+    for (const auto &partitionData : receiveBuffer[partition])
+      for (const auto &item : partitionData)
+        result.push_back(item);
+
+    // 6. Sort data according to global spectrum index
+    std::sort(
+        result.begin(), result.end(),
+        [](const std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data> &t1,
+           const std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data> &t2)
+            -> bool { return std::get<0>(t1) < std::get<0>(t2); });
+
```

Comment:
We now have vector of data in the correct order, i.e., the index into the vector is the spectrum index with the new partitioning. We could now use that vector to build the new instrument or `SpectrumInfo`.